src/main/java/org/dataone/speedbagit/SpeedBagIt.java

```diff
@@ -119,98 +178,108 @@ public void writeToTagManifest(String path, String checksum) {
+     * and bag.info are generated inside.
+     *
      * @param zos: The output stream that represents the streaming bag
-     * @throws IOException
+     * @throws IOException Throws when something went wrong with streaming the bag
+     * @throws NoSuchAlgorithmException Thrown when an unsupported checksum algorithm is used
      */
     public void stream(ZipOutputStream zos)
             throws IOException, NoSuchAlgorithmException {
-
-        logger.info("stream called!");
+        int totalSize = 0;
+        // Stream all of the files in the root 'data' directory
         for (SpeedFile streamingFile : this.dataFiles) {
-            logger.info(String.format("Preparing to write %s", streamingFile.getPath()));
-            ZipEntry entry = new ZipEntry(streamingFile.getPath());
-            zos.putNextEntry(entry);
-
-            SpeedStream fileStream = streamingFile.getStream();
-            fileStream.transferTo(zos);
+            this.streamFile(zos, streamingFile);
 
             String checksum = new String(streamingFile.getStream().getChecksum());
-            logger.info(checksum);
-            logger.info(String.valueOf(streamingFile.getStream().getSize()));
             this.writeToDataManifest(streamingFile.getPath(), checksum);
-            this.payloadOxum += streamingFile.getStream().getSize();
-            zos.closeEntry();
+            totalSize += streamingFile.getStream().getSize();
         }
+        String payloadOxum =  String.format("%s.%s",totalSize, this.dataFiles.size());
+        // Generate and add the bagit.txt file
+        InputStream bagTextStream = new ByteArrayInputStream(this.generateBagitTxt().getBytes(StandardCharsets.UTF_8));
+        this.addFile(bagTextStream, "bagit.txt", MessageDigest.getInstance(checksumAlgorithm), true);
 
-        // Now that payload-oxum is computed, add it to the bag-info.txt file
-        this.bagInfoFile = String.format("%spayload-oxum: %s\n", this.bagInfoFile, payloadOxum);
-        InputStream fileStream = new ByteArrayInputStream(this.bagInfoFile.getBytes(StandardCharsets.UTF_8));
-        this.addFile(fileStream, "bag-info.txt", MessageDigest.getInstance("MD5"), true);
+
+        // Generate and add the bag-info.txt file
+        String bagInfoFile = generateBagInfoTxt(payloadOxum, totalSize);
+        InputStream fileStream = new ByteArrayInputStream(bagInfoFile.getBytes(StandardCharsets.UTF_8));
+        this.addFile(fileStream, "bag-info.txt", MessageDigest.getInstance(checksumAlgorithm), true);
+
+        // Generate and add the data manifest file
         String fileName = String.format("manifest-%s.txt", this.checksumAlgorithm);
         fileStream = new ByteArrayInputStream(this.dataManifestFile.getBytes(StandardCharsets.UTF_8));
-        this.addFile(fileStream, fileName, MessageDigest.getInstance("MD5"), true);
+        this.addFile(fileStream, fileName, MessageDigest.getInstance(checksumAlgorithm), true);
```

Comment:
Same as above comment - any memory issues with `ByteArrayInputStream` for the data manifest file?

Reply:
The size of the data manifest file is O(# of files in the payload). I kept this in memory to avoid the overhead of writing to disk for every file in the payload (speed) even though it could theoretically overflow memory for insanely large bags (I would argue anyone making insanely big bags are neglecting BagIt's feature of splitting bags though).